pkg/networkpath/traceroute/tcp/utils_unix.go

```diff
@@ -44,75 +52,67 @@ func sendPacket(rawConn rawConnWrapper, header *ipv4.Header, payload []byte) err
 // receives a matching packet within the timeout, a blank response is returned.
 // Once a matching packet is received by a listener, it will cause the other listener
 // to be canceled, and data from the matching packet will be returned to the caller
-func listenPackets(icmpConn rawConnWrapper, tcpConn rawConnWrapper, timeout time.Duration, localIP net.IP, localPort uint16, remoteIP net.IP, remotePort uint16, seqNum uint32) (net.IP, uint16, layers.ICMPv4TypeCode, time.Time, error) {
-	var tcpErr error
-	var icmpErr error
-	var wg sync.WaitGroup
-	var icmpIP net.IP
-	var tcpIP net.IP
-	var icmpCode layers.ICMPv4TypeCode
-	var tcpFinished time.Time
-	var icmpFinished time.Time
-	var port uint16
-	wg.Add(2)
+func listenPackets(icmpConn rawConnWrapper, tcpConn rawConnWrapper, timeout time.Duration, localIP net.IP, localPort uint16, remoteIP net.IP, remotePort uint16, seqNum uint32) packetResponse {
+	respChan := make(chan packetResponse, 2)
 	ctx, cancel := context.WithTimeout(context.Background(), timeout)
 	defer cancel()
 	go func() {
-		defer wg.Done()
-		defer cancel()
-		tcpIP, port, _, tcpFinished, tcpErr = handlePackets(ctx, tcpConn, "tcp", localIP, localPort, remoteIP, remotePort, seqNum)
+		respChan <- handlePackets(ctx, tcpConn, localIP, localPort, remoteIP, remotePort, seqNum)
 	}()
 	go func() {
-		defer wg.Done()
-		defer cancel()
-		icmpIP, _, icmpCode, icmpFinished, icmpErr = handlePackets(ctx, icmpConn, "icmp", localIP, localPort, remoteIP, remotePort, seqNum)
+		respChan <- handlePackets(ctx, icmpConn, localIP, localPort, remoteIP, remotePort, seqNum)
 	}()
-	wg.Wait()
 
-	if tcpErr != nil && icmpErr != nil {
-		_, tcpCanceled := tcpErr.(common.CanceledError)
-		_, icmpCanceled := icmpErr.(common.CanceledError)
-		if icmpCanceled && tcpCanceled {
+	// wait for both responses to return
+	// as one could error even if the other
+	// succeeds
+	var err error
+	for i := 0; i < 2; i++ {
+		select {
+		case <-ctx.Done():
 			log.Trace("timed out waiting for responses")
-			return net.IP{}, 0, 0, time.Time{}, nil
-		}
-		if tcpErr != nil {
-			log.Errorf("TCP listener error: %s", tcpErr.Error())
-		}
-		if icmpErr != nil {
-			log.Errorf("ICMP listener error: %s", icmpErr.Error())
-		}
+			return packetResponse{
+				Err: err,
+			}
+		case resp := <-respChan:
+			if resp.Err == nil {
+				return resp
```

Comment:
So if we have a pending response, we just ignore it? No way to cancel listening for it?

Reply:
We `defer cancel()` when we create the `context.WithTimeout` above so this should cancel the context that gets passed in to each of these listeners. As soon as the remaining listener is back in their `select`, it will return. In the worst case scenario it would be an extra 100ms via `conn.ReadTimeout` assuming it just went back to reading.